### PR TITLE
rocAL Tensor Retinanet changes

### PR DESCRIPTION
**Pull request by fiona-gladwin**
- Add BOX IOU matcher implementation

### CODE REVIEW

File: rocAL/rocAL_pybind/amd/rocal/readers.py [hunk 1]
```diff
@@ -36,7 +36,8 @@ def coco(annotations_file='', ltrb=True, masks=False, ratio=False,
         "ltrb": ltrb,
         "is_box_encoder": is_box_encoder,
         "avoid_class_remapping": avoid_class_remapping,
-        "aspect_ratio_grouping": aspect_ratio_grouping}
+        "aspect_ratio_grouping": aspect_ratio_grouping,
+        "is_box_iou_matcher":is_box_iou_matcher}
```
Comment (fiona-gladwin) [inline rocAL/rocAL_pybind/amd/rocal/readers.py:40]: Space after :

File: rocAL/rocAL/include/pipeline/master_graph.h [hunk 2]
```diff
@@ -49,6 +49,7 @@ THE SOFTWARE.
 #define BBOX_COUNT 4
 #define MAX_NUM_ANCHORS 8732  // Num of bbox achors used in SSD training
 #define MAX_MASK_BUFFER 10000
+#define MAX_ANCHORS 120087  // Num of bbox achors used in Retinanet training
```
Comment (swetha097) [inline rocAL/rocAL/include/pipeline/master_graph.h:52]: Rename MAX_ANCHORS to MAX_RETINANET_ANCHORS

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 3]
```diff
@@ -1516,6 +1552,20 @@ TensorList * MasterGraph::mask_meta_data()
     return &_mask_tensor_list;
 }
 
+TensorList * MasterGraph::matches_meta_data()
+{
+    if(_ring_buffer.level() == 0)
+        THROW("No meta data has been loaded")
+    auto meta_data_buffers = (unsigned char *)(_ring_buffer.get_meta_read_buffers()[2]); // Get matches buffer from ring buffer
```
Comment (swetha097) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:1559]: Use static_cast

**Review by swetha097 — COMMENTED** [review]
Please address the first round of review comments

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 4]
```diff
@@ -1516,6 +1552,20 @@ TensorList * MasterGraph::mask_meta_data()
...
+    for(unsigned i = 0; i < _matches_tensor_list.size(); i++)
+    {
+        _matches_tensor_list[i]->set_mem_handle((void *)meta_data_buffers);
```
Comment (swetha097) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:1562]: Make use of static_cast

File: rocAL/rocAL_pybind/rocal_pybind.cpp [hunk 5]
```diff
@@ -708,17 +708,16 @@ namespace rocal{
...
-        //                     sizeof(int),
-        //                     py::format_descriptor<int>::format(),
-        //                     1,
-        //                     {matches->size() * 120087},
-        //                     {sizeof(int) }));
-        // }, py::return_value_policy::reference);
+        m.def("getMatchedIndices", [](RocalContext context) {
+            rocalTensorList *matches = rocalGetMatchedIndices(context);
+            return py::array(py::buffer_info(
+                            (int *)(matches->at(0)->buffer()),
```
Comment (swetha097) [inline rocAL/rocAL_pybind/rocal_pybind.cpp:714]: Please use static_cast here

File: rocAL/rocAL/include/meta_data/bounding_box_graph.h [hunk 6]
```diff
@@ -26,8 +26,6 @@ THE SOFTWARE.
 #include "meta_node.h"
 
 typedef  struct { float xc; float yc; float w; float h; } BoundingBoxCord_xcycwh;
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply rocAL/rocAL/include/meta_data/bounding_box_graph.h:28]: It is handled in meta_data.h
Previously we had a struct with double bbox in metadata.h so we introduced the struct with float bbox here as it was required by the box encoder

File: rocAL/rocAL/include/pipeline/master_graph.h [hunk 7]
```diff
@@ -101,18 +102,21 @@ class MasterGraph
...
     std::vector<rocalTensorList *> create_label_reader(const char *source_path, MetaDataReaderType reader_type);
     std::vector<rocalTensorList *> create_video_label_reader(const char *source_path, MetaDataReaderType reader_type, unsigned sequence_length, unsigned frame_step, unsigned frame_stride, bool file_list_frame_num = true);
-    std::vector<rocalTensorList *> create_coco_meta_data_reader(const char *source_path, bool is_output, MetaDataReaderType reader_type, MetaDataType label_type, bool ltrb_bbox = true, bool is_box_encoder = false, bool avoid_class_remapping=false, bool aspect_ratio_grouping = false, float sigma = 0.0, unsigned pose_output_width = 0, unsigned pose_output_height = 0);
+    std::vector<rocalTensorList *> create_coco_meta_data_reader(const char *source_path, bool is_output, MetaDataReaderType reader_type, MetaDataType label_type, bool ltrb_bbox = true, bool is_box_encoder = false,
+                                                                bool avoid_class_remapping = false, bool avoid_class_remapping = false, bool is_box_iou_matcher = false, float sigma = 0.0, unsigned pose_output_width = 0, unsigned pose_output_height = 0);
```
Comment (SundarRajan28) [inline rocAL/rocAL/include/pipeline/master_graph.h:106]: avoid_class_remapping is repeated twice. Chnage it to aspect_ratio_grouping

File: rocAL/rocAL/include/api/rocal_api_meta_data.h [hunk 8]
```diff
@@ -198,11 +198,27 @@ extern "C" RocalMetaData ROCAL_API_CALL rocalGetEncodedBoxesAndLables(RocalConte
 ///
 /// \param rocal_context
 /// \param buf The user's buffer that will be filled with image id info for the images in the output batch.
-extern "C" void ROCAL_API_CALL rocalGetImageId(RocalContext p_context,  int* buf);
+extern "C" void ROCAL_API_CALL rocalGetImageId(RocalContext p_context, int* buf);
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply rocAL/rocAL/include/api/rocal_api_meta_data.h:218]: It is added here?

File: rocAL/rocAL/source/meta_data/bounding_box_graph.cpp [hunk 9]
```diff
@@ -232,14 +232,81 @@ void BoundingBoxGraph::update_box_encoder_meta_data(std::vector<float> *anchors,
...
+        std::vector<int> low_quality_preds(anchors_size, -1);
+
+        // Calculate IoU's, The number of IoU Values calculated will be (bb_count x anchors_size)
+        for (unsigned bb_idx = 0; bb_idx < bb_count; bb_idx++) {
+            BoundingBoxCord box = bb_coords[bb_idx];
+            float box_area = (box.b - box.t) * (box.r - box.l);
+            float best_bbox_iou = -1.0f;
+            std::vector<float> bbox_iou(anchors_size);  // IoU value for bbox mapped with each anchor
+            for (unsigned int anchor_idx = 0; anchor_idx < anchors_size; anchor_idx++) {
+                float iou_val = ssd_BBoxIntersectionOverUnion(box, box_area, bbox_anchors[anchor_idx]);
```
Comment (swetha097) [inline rocAL/rocAL/source/meta_data/bounding_box_graph.cpp:275]: This naming of the functiom seems to contain both snake and camel case.
Change to snake_case

File: rocAL/rocAL/source/meta_data/bounding_box_graph.cpp [hunk 10]
```diff
@@ -232,14 +232,81 @@ void BoundingBoxGraph::update_box_encoder_meta_data(std::vector<float> *anchors,
...
+                    encoded_bb[anchor_idx].h = (-p_anchor->t + p_anchor->b);      //h
                     encoded_labels[anchor_idx] = 0;
                 }
             }
         }
     }
 }
 
+void BoundingBoxGraph::update_box_iou_matcher(
+    std::vector<float> *anchors, int *matches_idx_buffer,
```
Comment (swetha097) [inline rocAL/rocAL/source/meta_data/bounding_box_graph.cpp:247]: Align the args correctly

Comment (swetha097) [reply rocAL/rocAL/source/meta_data/bounding_box_graph.cpp:247]: done

File: rocAL/rocAL/source/meta_data/bounding_box_graph.cpp (hunk 9 above)
Comment (swetha097) [reply rocAL/rocAL/source/meta_data/bounding_box_graph.cpp]: done

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 11]
```diff
@@ -1508,6 +1544,20 @@ TensorList * MasterGraph::mask_meta_data()
...
+{
+    if(_ring_buffer.level() == 0)
+        THROW("No meta data has been loaded")
+    auto meta_data_buffers = reinterpret_cast<unsigned char *>(_ring_buffer.get_meta_read_buffers()[2]); // Get matches buffer from ring buffer
+    for(unsigned i = 0; i < _matches_tensor_list.size(); i++)
+    {
+        _matches_tensor_list[i]->set_mem_handle(reinterpret_cast<void *>(meta_data_buffers));
+        meta_data_buffers += _matches_tensor_list[i]->info().data_size();
+    }
+
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/source/pipeline/master_graph.cpp:1557]: done

File: rocAL/rocAL/source/meta_data/bounding_box_graph.cpp [hunk 12]
```diff
@@ -232,14 +232,81 @@ void BoundingBoxGraph::update_box_encoder_meta_data(std::vector<float> *anchors,
...
+    pMetaDataBatch full_batch_meta_data, float criteria, float high_threshold,
+    float low_threshold, bool allow_low_quality_matches) 
+{
+    auto bb_coords_batch = full_batch_meta_data->get_bb_cords_batch();
+    unsigned anchors_size = anchors->size() / 4;  // divide the anchors_size by 4 to get the total number of anchors
+    BoundingBoxCord *bbox_anchors = reinterpret_cast<BoundingBoxCord *>(anchors->data());
+
+    std::vector<int *> matches(full_batch_meta_data->size());
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/source/meta_data/bounding_box_graph.cpp:254]: Yes

**Pull request closed without merging**